lib/chef/data_collector.rb

```diff
@@ -0,0 +1,277 @@
+          update_current_resource_report(nil)
+        end
+      end
+
+      def run_list_expanded(run_list_expansion)
+        @expanded_run_list = run_list_expansion
+      end
+
+      def run_list_expand_failed(node, exception)
+        update_error_description(
+          Formatters::ErrorMapper.run_list_expand_failed(
+            node,
+            exception
+          ).for_json
+        )
+      end
+
+      def cookbook_resolution_failed(expanded_run_list, exception)
+        update_error_description(
+          Formatters::ErrorMapper.cookbook_resolution_failed(
+            expanded_run_list,
+            exception
+          ).for_json
+        )
+      end
+
+      def cookbook_sync_failed(cookbooks, exception)
+        update_error_description(
+          Formatters::ErrorMapper.cookbook_sync_failed(
+            cookbooks,
+            exception
+          ).for_json
+        )
+      end
+
+      private
+
+      def disable_reporter_on_error
+        yield
+      rescue Timeout::Error, Errno::EINVAL, Errno::ECONNRESET, Errno::ECONNREFUSED, EOFError, Net::HTTPBadResponse, Net::HTTPHeaderSyntaxError, Net::ProtocolError => e
+        disable_data_collector_reporter
+        code = if e.respond_to?(:response) && e.response.code
+                 e.response.code.to_s
+               else
+                 "Exception Code Empty"
+               end
+
+        msg = "Error while reporting run start to Data Collector. " \
+              "URL: #{data_collector_server_url} " \
+              "Exception: #{code} -- #{e.message} "
+
+        if Chef::Config[:data_collector_raise_on_failure]
+          Chef::Log.error(msg)
+          raise
+        else
+          Chef::Log.warn(msg)
+        end
+      end
+
+      def send_to_data_collector(message)
+        return unless data_collector_accessible?
+
+        Chef::Log.debug("data_collector_reporter: POSTing the following message to #{data_collector_server_url}: #{message.to_json}")
+        http.post(nil, message.to_json, headers)
+      end
+
+      def send_run_completion
+        # If run_status is nil we probably failed before the client triggered
+        # the run_started callback. In this case we'll skip updating because
+        # we have nothing to report.
+        return unless run_status
+
+        send_to_data_collector(Serializers::NodeUpdate.new(run_status))
+        send_to_data_collector(
+          Serializers::RunEnd.new(
+            run_status: run_status,
+            expanded_run_list: expanded_run_list,
+            total_resource_count: resource_count,
+            updated_resources: updated_resources,
+            status: status,
+            error_descriptions: error_descriptions
+          )
+        )
+      end
+
+      def headers(additional_headers = {})
+        headers = { "Content-Type" => "application/json" }
+        headers["x-data-collector-token"] = data_collector_token unless data_collector_token.nil?
```

Comment:
There was some discussion about this in the PR for the RFC where Adam mentioned:

> this should be a shared secret that gets hashed based on the payload. 

There was also the sense that we should version the protocol somehow.  I'm personally fine with the auth scheme being figured out in follow up PRs, but if we think we are gonna ship a version with it in the current state we might want to emit a warning to anyone who sets it that the auth is still in flux.